providers/rpc/logging.go

```diff
@@ -0,0 +1,61 @@
+package rpc
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+)
+
+type requestDetails struct {
+	Body    RequestPayload `json:"body"`
+	Headers http.Header    `json:"headers"`
+	URL     string         `json:"url"`
+	Method  string         `json:"method"`
+}
+
+type responseDetails struct {
+	StatusCode int             `json:"statusCode"`
+	Body       ResponsePayload `json:"body"`
+	Headers    http.Header     `json:"headers"`
+}
+
+// requestKVS returns a slice of key, value sets. Used for logging.
+func requestKVS(req *http.Request) []interface{} {
+	var r requestDetails
+	if req != nil && req.Body != nil {
+		var p RequestPayload
+		reqBody, err := io.ReadAll(req.Body)
+		if err == nil {
+			req.Body = io.NopCloser(bytes.NewBuffer(reqBody))
+			_ = json.Unmarshal(reqBody, &p)
+		}
```

Comment:
I believe there's a leak here because the underlying body is never closed on success, its just switched out with the `io.NopCloser()`.

Reply:
To elaborate. This code reads the body on l.28, then if no error occurs proceeds to swap the body with a `NopCloser()` without closing the `req.Body`'s original `io.ReadCloser`. Consequently, we risk a leak, or in the least inhibit the underlying TCP connection from being reused as per the `http.Request` docs.

I mention the leak because I've observed not calling `req.Body.Close()` result in OS file descriptor (there's even articles dedicated to the problem https://www.j4mcs.dev/posts/golang-response-body/).

Reply:
🤦🏻 I've just realized by brain fart, this is a Request, not a Reponse ...